File: leetcode/facebook/range_sum_of_bst.py
from data_structures.trees import TreeNode

from typing import Optional


class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        if not root:
            return 0

        ts = 0

        if low <= root.val <= high:
            ts += root.val

        if low < root.val:
            ts += self.rangeSumBST(root.left, low, high)

        if high > root.val:
            ts += self.rangeSumBST(root.right, low, high)

        return ts

File: leetdotrs/src/facebook/range_sum_of_bst.rs
#![allow(dead_code)]

struct Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32 {
        if let Some(node) = root {
            let mut sum: i32 = 0;

            let current = node.borrow().val;

            if low <= current && current <= high {
                sum += current;
            }

            if low < current {
                sum += Solution::range_sum_bst(node.borrow().left.clone(), low, high);
            }

            if high > current {
                sum += Solution::range_sum_bst(node.borrow().right.clone(), low, high);
            } 

            return sum
        }

        0
    }
}

